File: packages/java-cas-client/src/main/java/com/cassystem/client/CasClient.java
package com.cassystem.client;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import okhttp3.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

/**
 * CAS SSO Client for Java applications.
 *
 * <pre>
 * CasConfig config = new CasConfig("https://your-cas-server.com", "client_id", "client_secret");
 * CasClient cas = new CasClient(config);
 *
 * // Generate login URL
 * String loginUrl = cas.getLoginUrl("https://your-app.com/dashboard");
 *
 * // Validate token
 * Map&lt;String, Object&gt; user = cas.validateToken(token);
 *
 * // Check roles
 * boolean isAdmin = cas.userHasRole(user, "admin");
 * </pre>
 */
public class CasClient {
    private static final Logger logger = LoggerFactory.getLogger(CasClient.class);
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
    private static final Gson gson = new Gson();

    private final CasConfig config;
    private final OkHttpClient httpClient;
    private final Map<String, CacheEntry> cache = new ConcurrentHashMap<>();

    public CasClient(CasConfig config) {
        this.config = config;
        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(config.getTimeoutSeconds(), TimeUnit.SECONDS)
                .readTimeout(config.getTimeoutSeconds(), TimeUnit.SECONDS)
                .build();
    }

    /**
     * Generate CAS SSO login URL.
     */
    public String getLoginUrl(String returnUrl) {
        String redirectUri = (returnUrl != null && !returnUrl.isEmpty()) ? returnUrl : config.getCallbackUrl();
        try {
            return config.getServerUrl() + "/sso/login?"
                    + "client_id=" + URLEncoder.encode(config.getClientId(), "UTF-8")
                    + "&response_type=token"
                    + "&redirect_uri=" + URLEncoder.encode(redirectUri, "UTF-8");
        } catch (Exception e) {
            throw new RuntimeException("Failed to encode URL", e);
        }
    }

    /**
     * Generate SSO token for a user via client credentials.
     *
     * @return Token data map or null
     */
    public Map<String, Object> generateSSOToken(String username) {
        try {
            long timestamp = System.currentTimeMillis() / 1000;
            Map<String, String> requestData = new LinkedHashMap<>();
            requestData.put("client_id", config.getClientId());
            requestData.put("client_secret", config.getClientSecret());
            requestData.put("username", username);

            Request.Builder builder = new Request.Builder()
                    .url(config.getServerUrl() + "/api/sso/token")
                    .post(RequestBody.create(gson.toJson(requestData), JSON))
                    .header("Content-Type", "application/json")
                    .header("X-Client-ID", config.getClientId())
                    .header("X-Timestamp", String.valueOf(timestamp));

            if (config.isEnableSignatureValidation()) {
                builder.header("X-Signature", generateSignature("POST", "/api/sso/token", requestData, timestamp));
            }

            try (Response response = httpClient.newCall(builder.build()).execute()) {
                if (response.isSuccessful() && response.body() != null) {
                    String body = response.body().string();
                    @SuppressWarnings("unchecked")
                    Map<String, Object> data = gson.fromJson(body, Map.class);
                    if (data.containsKey("token")) return data;
                }
            }
            return null;
        } catch (Exception e) {
            logger.error("CAS SSO token generation failed: {}", e.getMessage());
            return null;
        }
    }

    /**
     * Validate SSO token with CAS server.
     *
     * @return User data map or null
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> validateToken(String token) {
        try {
            long timestamp = System.currentTimeMillis() / 1000;
            Map<String, String> requestData = new LinkedHashMap<>();
            requestData.put("token", token);
            requestData.put("client_id", config.getClientId());
            requestData.put("client_secret", config.getClientSecret());

            Request.Builder builder = new Request.Builder()
                    .url(config.getServerUrl() + "/api/sso/validate")
                    .post(RequestBody.create(gson.toJson(requestData), JSON))
                    .header("Content-Type", "application/json")
                    .header("X-Client-ID", config.getClientId())
                    .header("X-Timestamp", String.valueOf(timestamp));

            if (config.isEnableSignatureValidation()) {
                builder.header("X-Signature", generateSignature("POST", "/api/sso/validate", requestData, timestamp));
            }

            try (Response response = httpClient.newCall(builder.build()).execute()) {
                if (response.isSuccessful() && response.body() != null) {
                    String body = response.body().string();
                    Map<String, Object> data = gson.fromJson(body, Map.class);
                    if (data.containsKey("user")) {
                        Map<String, Object> user = (Map<String, Object>) data.get("user");
                        String cacheKey = md5(token);
                        cache.put(cacheKey, new CacheEntry(user, System.currentTimeMillis() + 3600_000));
                        return user;
                    }
                }
            }
            return null;
        } catch (Exception e) {
            logger.error("CAS token validation failed: {}", e.getMessage());
            return null;
        }
    }

    /** Get cached user data from token. */
    public Map<String, Object> getUserFromToken(String token) {
        String cacheKey = md5(token);
        CacheEntry entry = cache.get(cacheKey);
        if (entry != null && entry.expires > System.currentTimeMillis()) {
            return entry.data;
        }
        cache.remove(cacheKey);
        return null;
    }

    /** Logout from CAS server. */
    public boolean logout(String token) {
        try {
            if (token != null) cache.remove(md5(token));
            Request request = new Request.Builder()
                    .url(config.getServerUrl() + "/api/logout")
                    .post(RequestBody.create("{}", JSON))
                    .build();
            try (Response response = httpClient.newCall(request).execute()) {
                return response.isSuccessful();
            }
        } catch (Exception e) {
            logger.error("CAS logout failed: {}", e.getMessage());
            return false;
        }
    }

    /** Check if user has a specific role. */
    @SuppressWarnings("unchecked")
    public boolean userHasRole(Map<String, Object> user, String role) {
        List<String> roles = (List<String>) user.getOrDefault("roles", Collections.emptyList());
        return roles.contains(role);
    }

    /** Check if user has any of the specified roles. */
    @SuppressWarnings("unchecked")
    public boolean userHasAnyRole(Map<String, Object> user, String... roles) {
        List<String> userRoles = (List<String>) user.getOrDefault("roles", Collections.emptyList());
        for (String role : roles) {
            if (userRoles.contains(role)) return true;
        }
        return false;
    }

    /** Check if user has all specified roles. */
    @SuppressWarnings("unchecked")
    public boolean userHasAllRoles(Map<String, Object> user, String... roles) {
        List<String> userRoles = (List<String>) user.getOrDefault("roles", Collections.emptyList());
        for (String role : roles) {
            if (!userRoles.contains(role)) return false;
        }
        return true;
    }

    // --- Internal helpers ---

    private String generateSignature(String method, String uri, Map<String, String> data, long timestamp) {
        try {
            String body = gson.toJson(data);
            String payload = String.join("|", method, uri, body, String.valueOf(timestamp), config.getClientId());
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(config.getSignatureSecret().getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            byte[] hash = mac.doFinal(payload.getBytes(StandardCharsets.UTF_8));
            return "sha256=" + bytesToHex(hash);
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate signature", e);
        }
    }

    private static String md5(String input) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            return bytesToHex(md.digest(input.getBytes(StandardCharsets.UTF_8)));
        } catch (Exception e) {
            return input;
        }
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) sb.append(String.format("%02x", b));
        return sb.toString();
    }

    private static class CacheEntry {
        final Map<String, Object> data;
        final long expires;
        CacheEntry(Map<String, Object> data, long expires) {
            this.data = data;
            this.expires = expires;
        }
    }
}

File: packages/dotnet-cas-client/CasClient.cs
using System.Collections.Concurrent;
using System.Net.Http.Json;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Web;
using Microsoft.Extensions.Logging;

namespace CasSystem.Client;

/// <summary>
/// CAS SSO Client for .NET applications.
/// </summary>
/// <example>
/// var config = new CasConfig { ServerUrl = "https://your-cas-server.com", ClientId = "id", ClientSecret = "secret" };
/// var cas = new CasClient(config);
/// var user = await cas.ValidateTokenAsync(token);
/// </example>
public class CasClient
{
    private readonly CasConfig _config;
    private readonly HttpClient _http;
    private readonly ILogger<CasClient>? _logger;
    private readonly ConcurrentDictionary<string, CacheEntry> _cache = new();

    public CasClient(CasConfig config, ILogger<CasClient>? logger = null)
    {
        _config = config ?? throw new ArgumentNullException(nameof(config));
        _logger = logger;

        if (string.IsNullOrEmpty(config.ServerUrl)) throw new ArgumentException("ServerUrl is required");
        if (string.IsNullOrEmpty(config.ClientId)) throw new ArgumentException("ClientId is required");
        if (string.IsNullOrEmpty(config.ClientSecret)) throw new ArgumentException("ClientSecret is required");

        _http = new HttpClient
        {
            BaseAddress = new Uri(config.ServerUrl.TrimEnd('/')),
            Timeout = TimeSpan.FromSeconds(config.TimeoutSeconds)
        };
        _http.DefaultRequestHeaders.Add("Accept", "application/json");
    }

    /// <summary>Generate CAS SSO login URL.</summary>
    public string GetLoginUrl(string? returnUrl = null)
    {
        var redirectUri = !string.IsNullOrEmpty(returnUrl) ? returnUrl : _config.CallbackUrl;
        return $"{_config.ServerUrl.TrimEnd('/')}/sso/login?client_id={HttpUtility.UrlEncode(_config.ClientId)}&response_type=token&redirect_uri={HttpUtility.UrlEncode(redirectUri)}";
    }

    /// <summary>Generate SSO token for a user via client credentials.</summary>
    public async Task<Dictionary<string, object>?> GenerateSSOTokenAsync(string username)
    {
        try
        {
            var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            var requestData = new Dictionary<string, string>
            {
                ["client_id"] = _config.ClientId,
                ["client_secret"] = _config.ClientSecret,
                ["username"] = username
            };

            var request = CreateRequest(HttpMethod.Post, "/api/sso/token", requestData, timestamp);
            var response = await _http.SendAsync(request);

            if (response.IsSuccessStatusCode)
            {
                var data = await response.Content.ReadFromJsonAsync<Dictionary<string, object>>();
                if (data != null && data.ContainsKey("token")) return data;
            }
            return null;
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "CAS SSO token generation failed");
            return null;
        }
    }

    /// <summary>Validate SSO token with CAS server.</summary>
    public async Task<Dictionary<string, object>?> ValidateTokenAsync(string token)
    {
        try
        {
            var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            var requestData = new Dictionary<string, string>
            {
                ["token"] = token,
                ["client_id"] = _config.ClientId,
                ["client_secret"] = _config.ClientSecret
            };

            var request = CreateRequest(HttpMethod.Post, "/api/sso/validate", requestData, timestamp);
            var response = await _http.SendAsync(request);

            if (response.IsSuccessStatusCode)
            {
                var json = await response.Content.ReadAsStringAsync();
                using var doc = JsonDocument.Parse(json);
                if (doc.RootElement.TryGetProperty("user", out var userElem))
                {
                    var user = JsonSerializer.Deserialize<Dictionary<string, object>>(userElem.GetRawText());
                    if (user != null)
                    {
                        var cacheKey = ComputeMd5(token);
                        _cache[cacheKey] = new CacheEntry(user, DateTimeOffset.UtcNow.AddHours(1).ToUnixTimeMilliseconds());
                        return user;
                    }
                }
            }
            return null;
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "CAS token validation failed");
            return null;
        }
    }

    /// <summary>Get cached user data from token.</summary>
    public Dictionary<string, object>? GetUserFromToken(string token)
    {
        var key = ComputeMd5(token);
        if (_cache.TryGetValue(key, out var entry) && entry.Expires > DateTimeOffset.UtcNow.ToUnixTimeMilliseconds())
            return entry.Data;
        _cache.TryRemove(key, out _);
        return null;
    }

    /// <summary>Logout from CAS server.</summary>
    public async Task<bool> LogoutAsync(string? token = null)
    {
        try
        {
            if (token != null) _cache.TryRemove(ComputeMd5(token), out _);
            var response = await _http.PostAsync("/api/logout", new StringContent("{}", Encoding.UTF8, "application/json"));
            return response.IsSuccessStatusCode;
        }
        catch (Exception ex)
        {
            _logger?.LogError(ex, "CAS logout failed");
            return false;
        }
    }

    /// <summary>Check if user has a specific role.</summary>
    public static bool UserHasRole(Dictionary<string, object> user, string role)
    {
        if (user.TryGetValue("roles", out var rolesObj) && rolesObj is JsonElement elem)
            return elem.EnumerateArray().Any(r => r.GetString() == role);
        return false;
    }

    /// <summary>Check if user has any of the specified roles.</summary>
    public static bool UserHasAnyRole(Dictionary<string, object> user, params string[] roles)
    {
        if (user.TryGetValue("roles", out var rolesObj) && rolesObj is JsonElement elem)
        {
            var userRoles = elem.EnumerateArray().Select(r => r.GetString()).ToHashSet();
            return roles.Any(r => userRoles.Contains(r));
        }
        return false;
    }

    /// <summary>Check if user has all specified roles.</summary>
    public static bool UserHasAllRoles(Dictionary<string, object> user, params string[] roles)
    {
        if (user.TryGetValue("roles", out var rolesObj) && rolesObj is JsonElement elem)
        {
            var userRoles = elem.EnumerateArray().Select(r => r.GetString()).ToHashSet();
            return roles.All(r => userRoles.Contains(r));
        }
        return false;
    }

    // --- Internal ---

    private HttpRequestMessage CreateRequest(HttpMethod method, string uri, Dictionary<string, string> data, long timestamp)
    {
        var json = JsonSerializer.Serialize(data);
        var req = new HttpRequestMessage(method, uri)
        {
            Content = new StringContent(json, Encoding.UTF8, "application/json")
        };
        req.Headers.Add("X-Client-ID", _config.ClientId);
        req.Headers.Add("X-Timestamp", timestamp.ToString());

        if (_config.EnableSignatureValidation)
            req.Headers.Add("X-Signature", GenerateSignature(method.Method, uri, json, timestamp));

        return req;
    }

    private string GenerateSignature(string method, string uri, string body, long timestamp)
    {
        var payload = $"{method}|{uri}|{body}|{timestamp}|{_config.ClientId}";
        using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(_config.SignatureSecret));
        var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(payload));
        return "sha256=" + Convert.ToHexString(hash).ToLowerInvariant();
    }

    private static string ComputeMd5(string input)
    {
        var hash = MD5.HashData(Encoding.UTF8.GetBytes(input));
        return Convert.ToHexString(hash).ToLowerInvariant();
    }

    private record CacheEntry(Dictionary<string, object> Data, long Expires);
}
